state/snapshotsManager.go

```diff
@@ -0,0 +1,454 @@
+	}
+
+	log.Debug("startSnapshotAfterRestart", "rootHash", rootHash, "epoch", epoch)
+	sm.SnapshotState(rootHash, epoch, trieStorageManager)
+
+	return nil
+}
+
+func (sm *snapshotsManager) getSnapshotRootHashAndEpoch(trieStorageManager common.StorageManager) ([]byte, uint32, error) {
+	sm.mutex.RLock()
+	defer sm.mutex.RUnlock()
+
+	rootHash, err := trieStorageManager.GetFromCurrentEpoch([]byte(lastSnapshot))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	epoch, err := trieStorageManager.GetLatestStorageEpoch()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return rootHash, epoch, nil
+}
+
+// SnapshotState starts a state snapshot
+func (sm *snapshotsManager) SnapshotState(
+	rootHash []byte,
+	epoch uint32,
+	trieStorageManager common.StorageManager,
+) {
+	sm.mutex.Lock()
+
+	stats, skipSnapshot := sm.verifyAndStartSnapshot(rootHash, epoch, trieStorageManager)
+	if skipSnapshot {
+		log.Debug("skipping snapshot",
+			"last snapshot rootHash", sm.lastSnapshot.rootHash,
+			"rootHash", rootHash,
+			"last snapshot epoch", sm.lastSnapshot.epoch,
+			"epoch", epoch,
+			"isSnapshotInProgress", sm.isSnapshotInProgress.IsSet(),
+		)
+
+		sm.mutex.Unlock()
+		return
+	}
+	sm.mutex.Unlock()
+
+	log.Info("starting snapshot", "type", sm.stateMetrics.GetSnapshotMessage(), "rootHash", rootHash, "epoch", epoch)
+
+	go sm.snapshotState(rootHash, epoch, trieStorageManager, stats)
+
+	sm.waitForCompletionIfAppropriate(stats)
+}
+
+// SetStateCheckpoint sets a checkpoint for the state trie
+func (sm *snapshotsManager) SetStateCheckpoint(rootHash []byte, trieStorageManager common.StorageManager) {
+	sm.setStateCheckpoint(rootHash, trieStorageManager)
+}
+
+func (sm *snapshotsManager) setStateCheckpoint(rootHash []byte, trieStorageManager common.StorageManager) {
+	log.Trace("accountsDB.SetStateCheckpoint", "root hash", rootHash)
+	trieStorageManager.EnterPruningBufferingMode()
+
+	missingNodesChannel := make(chan []byte, missingNodesChannelSize)
+	iteratorChannels := sm.channelsProvider.GetIteratorChannels()
+
+	stats := newSnapshotStatistics(1, 1)
+	go func() {
+		stats.NewSnapshotStarted()
+		trieStorageManager.SetCheckpoint(rootHash, rootHash, iteratorChannels, missingNodesChannel, stats)
+		sm.snapshotUserAccountDataTrie(false, rootHash, iteratorChannels, missingNodesChannel, stats, 0, trieStorageManager)
+
+		stats.SnapshotFinished()
+	}()
+
+	go sm.syncMissingNodes(missingNodesChannel, iteratorChannels.ErrChan, stats, sm.getTrieSyncer())
+
+	// TODO decide if we need to take some actions whenever we hit an error that occurred in the checkpoint process
+	//  that will be present in the errChan var
+	go sm.finishSnapshotOperation(rootHash, stats, missingNodesChannel, "setStateCheckpoint"+sm.stateMetrics.GetSnapshotMessage(), trieStorageManager)
+
+	sm.waitForCompletionIfAppropriate(stats)
+}
+
+func (sm *snapshotsManager) verifyAndStartSnapshot(rootHash []byte, epoch uint32, trieStorageManager common.StorageManager) (*snapshotStatistics, bool) {
```

Comment:
is this a new function?

Reply:
No, is a rename of `prepareSnapshot()` with some small changes. Reverted the name as `prepareSnapshot` is more accurate to what the func actually does